teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+---
+status: proposed
+title: Array result types
+creation-date: '2021-07-14'
+last-updated: '2021-07-14'
+authors:
+- '@bobcatfish'
+---
+
+# TEP-0076: Expanded array support: results and indexing
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases](#use-cases)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats](#notescaveats)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Upgrade &amp; Migration Strategy](#upgrade--migration-strategy)
+- [Implementation Pull request(s)](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+This TEP proposes expanding support for arrays by adding:
+* Support for indexing into arrays in variable replacement
+  (in addition to [replacing entire lists](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#substituting-array-parameters))
+* Support for array [results](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results) (for both Pipelines and Tasks)
+
+## Motivation
+
+* See [use cases](#use-cases) - use cases for looping support in pipelines often need an array input;
+  being able to produce an array from a Task to feed into a loop would enable some interesting CD use cases
+* TEP-0075 proposing adding dictionary support in addition to the existing support strings
+  and arrays as params, and results, but today we only support arrays as params and not as
+  results. The proposal to add dictionary is mostly motivated by use cases for dictionary
+  results (vs params) - but today we only allow string results. It seems confusingly
+  inconsistent to imagine adding dictionary results but not array results (even though
+  we support array params)
+* Without indexing, arrays will only work with arrays. This means you cannot combine arrays with Tasks that
+  only take string parameters and this limits the interoperabiltiy between Tasks.
+
+### Goals
+
+* Add two missing pieces of array support
+  * As results
+  * Indexing into arrays
```

Comment:
I think there's no use case defined to support the indexing goal.
If we add indexing, we might need to introduce an operator that allows checking the size of an array too?

Reply:
Good point 😅

Reply:
another +1 to including the size as a variable replacement, that could be a way to solve https://github.com/tektoncd/pipeline/issues/4097

(will need to check if json schema provides something like this or if we would need to pave new ground)

Reply:
@bobcatfish This seems like a good use case to have in the TEP! 👍

Reply:
circling back to this, I forgot that @sbwsg and I had already discussed this a bit and I'm back to leaving this out of scope for now (discussion: https://github.com/tektoncd/community/pull/477#issuecomment-901308105 TL;DR not clear how to support this for other types e.g. objects, or where we'd draw the line - also we can always add this later!) - it's currently listed in the alternatives section of the doc